Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,332 @@
+
+            private var file: UnsafeMutablePointer<FILE>?
+
+            init(_ path: String) {
+                self.path = path
+            }
+
+            deinit {
+                assert(self.file == nil)
+            }
+
+            func open() {
+                guard let f = fopen(path, "r") else {
+                    return
+                }
+                self.file = f
+            }
+
+            func close() {
+                if let f = self.file {
+                    self.file = nil
+                    let success = fclose(f) == 0
+                    assert(success)
+                }
+            }
+
+            func readLine() -> String? {
+                guard let f = self.file else {
+                    return nil
+                }
+                let buff: [CChar] = Array(unsafeUninitializedCapacity: 1024) { ptr, size in
+                    guard fgets(ptr.baseAddress, Int32(ptr.count), f) != nil else {
+                        if feof(f) != 0 {
+                            size = 0
+                            return
+                        } else {
+                            preconditionFailure("Error reading line")
+                        }
+                    }
+                    size = strlen(ptr.baseAddress!)
+                }
+                if buff.isEmpty { return nil }
+                return String(cString: buff)
+            }
+
+            func readFull() -> String {
+                var s = ""
+                func loop() -> String {
+                    if let l = readLine() {
+                        s += l
+                        return loop()
+                    }
+                    return s
+                }
+                return loop()
+            }
+        }
+
+        let ticks = _SC_CLK_TCK
+
+        let file = CFile("/proc/self/stat")
+        file.open()
+        defer {
+            file.close()
+        }
+
+        guard
+            let statString = file.readFull()
+            .split(separator: ")")
+            .last
+        else { return nil }
+        let stats = String(statString)
+            .split(separator: " ")
+            .map(String.init)
+        guard
+            let virtualMemoryBytes = Int(stats[safe: 20]),
+            let rss = Int(stats[safe: 21]),
+            let startTimeTicks = Int(stats[safe: 19]),
+            let utimeTicks = Int(stats[safe: 11]),
+            let stimeTicks = Int(stats[safe: 12])
+        else { return nil }
+        let residentMemoryBytes = rss * _SC_PAGESIZE
+        let startTimeSeconds = startTimeTicks / ticks
+        let cpuSeconds = (utimeTicks / ticks) + (stimeTicks / ticks)
+
+        var _rlim = rlimit()
+
+        guard withUnsafeMutablePointer(to: &_rlim, { ptr in
+            getrlimit(__rlimit_resource_t(RLIMIT_NOFILE.rawValue), ptr) == 0
+        }) else { return nil }
+
+        let maxFileDescriptors = Int(_rlim.rlim_max)
+
+        guard let dir = opendir("/proc/self/fd") else { return nil }
```

Comment:
This can also be simplified with `FileManager`:

```swift
let openFileDescriptors = FileManager.default
    .contentsOfDirectory(atPath: "/proc/self/fd")
    .count
```